sw/device/lib/base/memory_hardened.h

```diff
@@ -0,0 +1,158 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_BASE_MEMORY_HARDENED_H_
+#define OPENTITAN_SW_DEVICE_LIB_BASE_MEMORY_HARDENED_H_
+
+/**
+ * @file
+ * @brief OpenTitan Hardened Memory Function Library
+ *
+ * This library provides hardened implementation of memory functions for
+ * copy, set and mask memory regions.
+ *
+ * Fast shuffling of memory accesses is done by choosing:
+ *  1. random stride for memory accesses
+ *  2. random starting position
+ *
+ * Stride is selected to be co-prime with length, so we can simply add it
+ * at each iteration and be sure we will go through all elements.
+ * This also gives opportunity to add decoy memory accesses.
+ *
+ * Such implementation avoids computing random permutations and still
+ * introduce enough noise, while being just 1.5x slower than regular memcpy().
+ */
+
+#include <stdalign.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/hardened.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif  // __cplusplus
+
+/**
+ * Fault Resistant memcpy(). Checks that copy was fully done.
+ * @param dest - destination buffer
+ * @param src - source buffer
+ * @param len - bytes to copy
+ * @return kHardenedBoolTrue if everything is ok, undefined value else.
+ *         The only guaranteed result is kHardenedBoolTrue.
+ */
+hardened_bool_t memcpy_fr(void *restrict dest, const void *restrict src,
```

Comment:
Please consider a more unified naming scheme for these functions. E.g. `fault_resistant_memcpy()` for this one.

Reply:
i renamed to memcpy_checked()